### PR TITLE
Add dynamic shape checks to nGraph transformations

### PR DESCRIPTION
**Pull request by GlebKazantaev**
In this PR I added checks for dynamic shapes for CommonOptimization/OpConversion transformations.

* PullTransposeThroughFQUp 
* BatchNormDecomposition 
* ConvertBroadcastToTiles 
* ConvertDepthToSpace 
* ConvertReduceToPooling 
* ConvertSpaceToDepth

### CODE REVIEW

File: inference-engine/src/transformations/src/transformations/common_optimizations/pull_transpose_through_fq.cpp [hunk 1]
```diff
@@ -9,56 +9,42 @@
...
-        if (!fq || fq->output(0).get_target_inputs().size() != 1) {
-            return false;
-        }
+    ngraph::matcher_pass_callback callback = [=](pattern::Matcher& m) {
+        auto & pattern_map = m.get_pattern_value_map();
+        auto transpose = pattern_map[m_transpose].get_node_shared_ptr();
+        auto fq = pattern_map[m_fq].get_node_shared_ptr();
 
-        auto input_shape = fq->input(0).get_source_output().get_shape();
+        auto input_shape = fq->input(0).get_partial_shape().rank().get_length();
```
Comment (ilyachur) [inline inference-engine/src/transformations/src/transformations/common_optimizations/pull_transpose_through_fq.cpp:30]: nit: it is not input shape, from my point of view input_rank is a better name.

File: inference-engine/src/transformations/src/transformations/common_optimizations/pull_transpose_through_fq.cpp [hunk 2]
```diff
@@ -9,56 +9,42 @@
...
         ngraph::OutputVector fq_inputs;
         for (size_t i = 0; i < fq->inputs().size(); ++i) {
-            std::shared_ptr<ngraph::Node> fq_input;
-            fq_input = fq->input(i).get_source_output().get_node_shared_ptr();
-            auto fq_input_shape = fq_input->get_shape();
+            auto fq_input = fq->input_value(i);
+            auto fq_input_rank = fq_input.get_partial_shape().rank().get_length();
             std::vector<int64_t> unsqueeze_axes;
-            for (size_t j = 0; j < input_shape.size() - fq_input_shape.size(); ++j) {
+            for (size_t j = 0; j < input_shape - fq_input_rank; ++j) {
```
Comment (ilyachur) [inline inference-engine/src/transformations/src/transformations/common_optimizations/pull_transpose_through_fq.cpp:38]: What if `input_shape` is less than `fq_input_rank`?

File: inference-engine/src/transformations/src/transformations/op_conversions/batch_norm_decomposition.cpp [hunk 3]
```diff
@@ -10,50 +10,45 @@
...
         // scale_add = variance + eps
         auto scale_add = make_shared<opset5::Add>(m_var, opset5::Constant::create(input_type, Shape{}, {m_bn->get_eps_value()}));
         // scale = sqrt(variance + eps)
         auto scale = make_shared<opset5::Sqrt>(scale_add);
         // Divide `gamma` by `sqrt(variance + eps)`
         auto gamma_div_scale = std::make_shared<opset5::Divide>(m_gamma, scale);
 
-        size_t dims_to_add = m_input->get_shape().size() - 2;
-        Shape input_aligned_shape = m_gamma->get_shape();
+        size_t dims_to_add = m_input.get_partial_shape().rank().get_length() - 2;
```
Comment (ilyachur) [inline inference-engine/src/transformations/src/transformations/op_conversions/batch_norm_decomposition.cpp:48]: What if rank is less than 2?

File: inference-engine/src/transformations/src/transformations/op_conversions/convert_broadcast_to_tiles.cpp [hunk 4]
```diff
@@ -23,17 +23,21 @@ ngraph::pass::ConvertBroadcastToTiles::ConvertBroadcastToTiles() {
             return false;
         }
 
-        auto data_node = broadcast->input_value(0).get_node_shared_ptr();
+        auto data_node = broadcast->input_value(0);
+        if (data_node.get_partial_shape().is_dynamic()) {
```
Comment (ilyachur) [inline inference-engine/src/transformations/src/transformations/op_conversions/convert_broadcast_to_tiles.cpp:27]: Why don't we use `pattern::any_input(pattern::has_static_shape()),` for this case?

File: inference-engine/src/transformations/src/transformations/common_optimizations/pull_transpose_through_fq.cpp (hunk 2 above)
Comment (GlebKazantaev) [reply inference-engine/src/transformations/src/transformations/common_optimizations/pull_transpose_through_fq.cpp:38]: rank().get_length() returns int64_t so the comparison with size_t will be ok in case of negative value

File: inference-engine/src/transformations/src/transformations/op_conversions/batch_norm_decomposition.cpp (hunk 3 above)
Comment (GlebKazantaev) [reply inference-engine/src/transformations/src/transformations/op_conversions/batch_norm_decomposition.cpp:48]: changed to int64_t

File: inference-engine/src/transformations/src/transformations/op_conversions/convert_broadcast_to_tiles.cpp (hunk 4 above)
Comment (GlebKazantaev) [reply inference-engine/src/transformations/src/transformations/op_conversions/convert_broadcast_to_tiles.cpp:27]: Broadcast has multiple inputs and there is no way to specify optional input with predicate. I'm working on that but for now I would keep this check here.

File: inference-engine/src/transformations/src/transformations/common_optimizations/pull_transpose_through_fq.cpp (hunk 1 above)
Comment (GlebKazantaev) [reply inference-engine/src/transformations/src/transformations/common_optimizations/pull_transpose_through_fq.cpp:30]: fixed.

File: inference-engine/src/transformations/src/transformations/common_optimizations/pull_transpose_through_fq.cpp (hunk 2 above)
Comment (ilyachur) [reply inference-engine/src/transformations/src/transformations/common_optimizations/pull_transpose_through_fq.cpp]: Ok, I thought that rank has a type uint64_t

File: inference-engine/src/transformations/src/transformations/op_conversions/batch_norm_decomposition.cpp (hunk 3 above)
Comment (ilyachur) [reply inference-engine/src/transformations/src/transformations/op_conversions/batch_norm_decomposition.cpp:48]: Sorry, I didn't see any changes 😄

File: inference-engine/src/transformations/src/transformations/op_conversions/batch_norm_decomposition.cpp [hunk 5]
```diff
@@ -84,45 +79,40 @@ ngraph::pass::BatchNormDecomposition::BatchNormDecomposition() {
...
         // scale_add = variance + eps
         auto scale_add = make_shared<opset5::Add>(m_var, opset5::Constant::create(input_type, Shape{}, {m_bn->get_eps_value()}));
         // scale = sqrt(variance + eps)
         auto scale = make_shared<opset5::Sqrt>(scale_add);
         // Divide `gamma` by `sqrt(variance + eps)`
         auto gamma_div_scale = std::make_shared<opset5::Divide>(m_gamma, scale);
 
-        size_t dims_to_add = m_input->get_shape().size() - 2;
-        Shape input_aligned_shape = m_gamma->get_shape();
+        int64_t dims_to_add = m_input.get_partial_shape().rank().get_length() - 2;
```
Comment (ilyachur) [inline inference-engine/src/transformations/src/transformations/op_conversions/batch_norm_decomposition.cpp:112]: But here you changed the type 👍

File: inference-engine/src/transformations/src/transformations/op_conversions/batch_norm_decomposition.cpp (hunk 3 above)
Comment (GlebKazantaev) [reply inference-engine/src/transformations/src/transformations/op_conversions/batch_norm_decomposition.cpp]: Changed.

File: inference-engine/src/transformations/src/transformations/op_conversions/batch_norm_decomposition.cpp (hunk 5 above)
Comment (GlebKazantaev) [reply inference-engine/src/transformations/src/transformations/op_conversions/batch_norm_decomposition.cpp:112]: yep :)

**Review by ilyachur — APPROVED** [review]
